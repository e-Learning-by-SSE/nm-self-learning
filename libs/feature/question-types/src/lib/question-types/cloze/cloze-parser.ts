export type Gap = {
	type: "C" | "T";
	values: {
		text: string;
		isCorrect: boolean;
	}[];
	startIndex: number;
	endIndex: number;
};

// {C: [#OptionWrong, OptionCorrect]}
// {T: [Text]}
// {T: [Text, Alternative]}

//Cloze regex catches all cloze gaps in the text
// eslint-disable-next-line no-useless-escape
const clozeRegex = /(?:{[TC]\: \[((?:#*(?:[A-Za-zÄÜÖäüö]+|(?:\$\$.+\$\$)),* *)+)\]})/gm;

//Latex regex detects if latex is used in the cloze gap
// eslint-disable-next-line no-useless-escape
const latexRegex = /\$\$[^\$]+\$\$/gm;
const GAP_PLACEHOLDER = "{GAP}";

export function parseCloze(text: string): Gap[] {
	const gaps: Gap[] = [];

	const lines = text.split("\n");
	const latexMap = new Map<string, string>();

	// checks line break style because they are different in windows and linux
	const lineBreakLength = text.includes("\r\n") ? 2 : 1;
	let lineIndex = 0;

	for (let i = 0; i < lines.length; i++) {
		//removes all line breaks
		lines[i] = lines[i].replace(/\r/g, "");
		lines[i] = lines[i].replace(/\n/g, "");

		const matches = [...lines[i].matchAll(clozeRegex)];
		if (matches && matches.length > 0) {
			for (const match of matches) {
				let valueString = match[1];
				const hasLatex = [...valueString.matchAll(latexRegex)];

				//replace latex with a random key to prevent the parser from parsing the latex
				for (let j = 0; j < hasLatex.length; j++) {
					const latex = hasLatex[j];
					const latexKey = "$" + j + "$";
					valueString = valueString.replace(latex[0], latexKey);
					latexMap.set(latexKey, latex[0]);
				}
				//gets the values of the gab
				const values: Gap["values"] = [];
				valueString.split(",").forEach(value => {
					const trimmedValue = value.trim();
					values.push({
						//remove the #
						text: replacePlaceholderWithLatex(
							trimmedValue.startsWith("#") ? trimmedValue.slice(1) : trimmedValue,
							latexMap
						),
						isCorrect: !trimmedValue.startsWith("#")
					});
				});

				//startindex for each line for the gap in the cloze text
				gaps.push({
					type: match[0].startsWith("{T") ? "T" : "C",
					values: values,
					startIndex: lineIndex + (match.index ?? 0),
					endIndex: lineIndex + (match.index ?? 0) + match[0].length - 1
				});
			}
		}
		// adds length of line breaks to index
		lineIndex += lines[i].length + lineBreakLength;
	}

	return gaps;
}

function replacePlaceholderWithLatex(text: string, latexMap: Map<string, string>) {
	return latexMap.get(text.trim()) ?? text;
}

export function insertPlaceholder(text: string, gaps: Gap[]): string {
	let index = 0;
	let result = "";

	for (const gap of gaps) {
		result += text.slice(index, gap.startIndex);
		result += GAP_PLACEHOLDER;
		index = gap.endIndex + 1;
	}

	result += text.slice(index);

	return result;
}

export function createCloze(text: string): { gaps: Gap[]; segments: string[] } {
	const gaps = parseCloze(text);
	const placeholderText = insertPlaceholder(text, gaps);
	return {
		gaps,
		segments: placeholderText.split(GAP_PLACEHOLDER)
	};
}
