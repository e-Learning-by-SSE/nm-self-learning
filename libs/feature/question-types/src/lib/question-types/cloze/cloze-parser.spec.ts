import { insertPlaceholder, parseCloze } from "./cloze-parser";

describe("Cloze Parser", () => {
	describe("parseCloze", () => {
		it("One sentence", () => {
			const gaps = parseCloze("Text before {C: [#a, b]} Text after");
			expect(gaps).toHaveLength(1);
			expect(gaps).toMatchInlineSnapshot(`
			Array [
			  Object {
			    "endIndex": 23,
			    "startIndex": 12,
			    "type": "C",
			    "values": Array [
			      Object {
			        "isCorrect": false,
			        "text": "a",
			      },
			      Object {
			        "isCorrect": true,
			        "text": "b",
			      },
			    ],
			  },
			]
		`);
		});

		it("Two sentences", () => {
			const gaps = parseCloze(
				"This is the {C: [#a, b]} first sentence. This is the {C: [c, #d]} second sentence."
			);
			expect(gaps).toHaveLength(2);
		});

		it("Two lines", () => {
			const gaps = parseCloze(
				"This is the {C: [#a, b]} first sentence.\nThis is the {C: [c, #d]} second sentence."
			);
			// Is the same as inline snapshot above only exported as a file
			expect(gaps).toMatchSnapshot();
		});

		it("Two lines but with whitespace in the first", () => {
			const gaps = parseCloze("  \n This is the {C: [c, #d]} second sentence.");
			// Is the same as inline snapshot above only exported as a file
			expect(gaps).toMatchSnapshot();
		});

		it("should handle latex correctly", () => {
			const text = "{T: [$$latex1$$, #$$latex2$$]}";
			const result = parseCloze(text);
			// Is the same as inline snapshot above only exported as a file
			expect(result).toMatchSnapshot();
		});

		it("should handle dashes in answer options correctly", () => {
			const text = "{C: [#a-b, c-d]}";
			const gaps = parseCloze(text);
			expect(gaps).toHaveLength(1);
		});
	});

	describe("insertPlaceholder", () => {
		it("{C: [#a, b]} -> {GAP}", () => {
			const text = "{C: [#a, b]}";
			const result = insertPlaceholder(text, parseCloze(text));
			expect(result).toEqual("{GAP}");
		});

		it("Before {C: [#a, b]} -> Before {GAP}", () => {
			const text = "ABC {C: [#a, b]}";
			const result = insertPlaceholder(text, parseCloze(text));
			expect(result).toMatchInlineSnapshot(`"ABC {GAP}"`);
		});

		it("{C: [#a, b]} After -> {GAP} After", () => {
			const text = "ABC {C: [#a, b]}";
			const result = insertPlaceholder(text, parseCloze(text));
			expect(result).toMatchInlineSnapshot(`"ABC {GAP}"`);
		});

		it("Front {C: [#a, b]} Middle {C: [#a, b]} Back -> Front {GAP} Middle {GAP} Back", () => {
			const text = "Front {C: [#a, b]} Middle {C: [#a, b]} Back";
			const result = insertPlaceholder(text, parseCloze(text));
			expect(result).toMatchInlineSnapshot(`"Front {GAP} Middle {GAP} Back"`);
		});
	});
});
