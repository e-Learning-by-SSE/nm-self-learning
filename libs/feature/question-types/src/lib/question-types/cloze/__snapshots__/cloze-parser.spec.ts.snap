// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cloze Parser parseCloze Two lines 1`] = `
Array [
  Object {
    "endIndex": 23,
    "startIndex": 12,
    "type": "C",
    "values": Array [
      Object {
        "isCorrect": false,
        "text": "a",
      },
      Object {
        "isCorrect": true,
        "text": "b",
      },
    ],
  },
  Object {
    "endIndex": 64,
    "startIndex": 53,
    "type": "C",
    "values": Array [
      Object {
        "isCorrect": true,
        "text": "c",
      },
      Object {
        "isCorrect": false,
        "text": "d",
      },
    ],
  },
]
`;

exports[`Cloze Parser parseCloze Two lines but with whitespace in the first 1`] = `
Array [
  Object {
    "endIndex": 27,
    "startIndex": 16,
    "type": "C",
    "values": Array [
      Object {
        "isCorrect": true,
        "text": "c",
      },
      Object {
        "isCorrect": false,
        "text": "d",
      },
    ],
  },
]
`;

exports[`Cloze Parser parseCloze should handle latex correctly 1`] = `
Array [
  Object {
    "endIndex": 29,
    "startIndex": 0,
    "type": "T",
    "values": Array [
      Object {
        "isCorrect": true,
        "text": "$$latex1$$",
      },
      Object {
        "isCorrect": false,
        "text": "$$latex2$$",
      },
    ],
  },
]
`;
